// react-instantsearch-dom-maps/dist/cjs/Control.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Control = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty3 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactInstantsearchDom = require("react-instantsearch-dom");

var _Provider = require("./Provider");

var _withGoogleMaps = _interopRequireDefault(require("./withGoogleMaps"));

var cx = (0, _reactInstantsearchDom.createClassNames)('GeoSearch');

var Control =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Control, _Component);

  function Control() {
    (0, _classCallCheck2.default)(this, Control);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Control).apply(this, arguments));
  }

  (0, _createClass2.default)(Control, [{
    key: "getStateContext",
    value: function getStateContext() {
      return this.context[_Provider.STATE_CONTEXT];
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          googleMapsInstance = _this$props.googleMapsInstance,
          translate = _this$props.translate;

      var _this$getStateContext = this.getStateContext(),
          isRefineOnMapMove = _this$getStateContext.isRefineOnMapMove,
          hasMapMoveSinceLastRefine = _this$getStateContext.hasMapMoveSinceLastRefine,
          toggleRefineOnMapMove = _this$getStateContext.toggleRefineOnMapMove,
          refineWithInstance = _this$getStateContext.refineWithInstance;

      return _react.default.createElement("div", {
        className: cx('control')
      }, isRefineOnMapMove || !hasMapMoveSinceLastRefine ? _react.default.createElement("label", {
        className: cx('label')
      }, _react.default.createElement("input", {
        className: cx('input'),
        type: "checkbox",
        checked: isRefineOnMapMove,
        onChange: toggleRefineOnMapMove
      }), translate('control')) : _react.default.createElement("button", {
        className: cx('redo'),
        onClick: function onClick() {
          return refineWithInstance(googleMapsInstance);
        }
      }, translate('redo')));
    }
  }]);
  return Control;
}(_react.Component);

exports.Control = Control;
(0, _defineProperty3.default)(Control, "propTypes", {
  googleMapsInstance: _propTypes.default.object.isRequired,
  translate: _propTypes.default.func.isRequired
});
(0, _defineProperty3.default)(Control, "contextTypes", (0, _defineProperty3.default)({}, _Provider.STATE_CONTEXT, _propTypes.default.shape({
  isRefineOnMapMove: _propTypes.default.bool.isRequired,
  toggleRefineOnMapMove: _propTypes.default.func.isRequired,
  hasMapMoveSinceLastRefine: _propTypes.default.bool.isRequired,
  refineWithInstance: _propTypes.default.func.isRequired
}).isRequired));

var _default = (0, _reactInstantsearchDom.translatable)({
  control: 'Search as I move the map',
  redo: 'Redo search here'
})((0, _withGoogleMaps.default)(Control));

exports.default = _default;
