// react-instantsearch/dist/native.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Configure", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchNative.Configure;
  }
});
Object.defineProperty(exports, "DynamicWidgets", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchNative.DynamicWidgets;
  }
});
Object.defineProperty(exports, "ExperimentalConfigureRelatedItems", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchNative.ExperimentalConfigureRelatedItems;
  }
});
Object.defineProperty(exports, "ExperimentalDynamicWidgets", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchNative.ExperimentalDynamicWidgets;
  }
});
Object.defineProperty(exports, "Index", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchNative.Index;
  }
});
Object.defineProperty(exports, "InstantSearch", {
  enumerable: true,
  get: function get() {
    return _reactInstantsearchNative.InstantSearch;
  }
});
var _reactInstantsearchNative = require("react-instantsearch-native");
